File: tasks/bench-arrow-fn-values/fixtures/reference/arrow.mjs
// Independent second implementation: tree-walking evaluator over an s-expression reader.
import { readFileSync } from 'node:fs';
export function runArrow(src) {
  const lines = src.split('\n').map((l) => l.trim()).filter((l) => l && !l.startsWith('#'));
  const out = [];
  const globals = new Map();
  const readTok = (s) => s.replace(/\(/g, ' ( ').replace(/\)/g, ' ) ').trim().split(/\s+/);
  const sexp = (t) => {
    let i = 0;
    const go = () => {
      const x = t[i++];
      if (x !== '(') return /^-?\d+$/.test(x) ? BigInt(x) : x;
      const l = [];
      while (t[i] !== ')') l.push(go());
      i++;
      return l;
    };
    return go();
  };
  const parseExpr = (s) => sexp(readTok(s));
  function parseBlock(i, stops) {
    const body = [];
    while (i < lines.length && !stops.includes(lines[i].split(' ')[0])) {
      const [st, j] = parseStmt(i);
      body.push(st);
      i = j;
    }
    return [body, i];
  }
  function parseStmt(i) {
    const line = lines[i];
    const sp = line.indexOf(' ');
    const head = sp < 0 ? line : line.slice(0, sp);
    const rest = sp < 0 ? '' : line.slice(sp + 1);
    switch (head) {
      case 'SET': { const q = rest.indexOf(' '); return [{ t: 'SET', n: rest.slice(0, q), e: parseExpr(rest.slice(q + 1)) }, i + 1]; }
      case 'PRINT': return [{ t: 'PRINT', e: parseExpr(rest) }, i + 1];
      case 'RET': return [{ t: 'RET', e: parseExpr(rest) }, i + 1];
      case 'GLOBAL': return [{ t: 'GLOBAL', n: rest }, i + 1];
      case 'IF': { const [a, j] = parseBlock(i + 1, ['ELSE', 'END']); let b = [], k = j; if (lines[j] === 'ELSE') [b, k] = parseBlock(j + 1, ['END']); return [{ t: 'IF', e: parseExpr(rest), a, b }, k + 1]; }
      case 'REPEAT': { const [a, j] = parseBlock(i + 1, ['END']); return [{ t: 'REPEAT', e: parseExpr(rest), a }, j + 1]; }
      case 'DEF': { const w = rest.split(' '); const [a, j] = parseBlock(i + 1, ['END']); return [{ t: 'DEF', n: w[0], ps: w.slice(1), a }, j + 1]; }
    }
    throw new Error('bad line ' + line);
  }
  const [prog] = parseBlock(0, []);
  class Ret { constructor(v) { this.v = v; } }
  const fnv = (ps, body, isProc) => ({ ps, body, isProc });
  function get(n, fr) {
    if (fr && !fr.gl.has(n) && fr.loc.has(n)) return fr.loc.get(n);
    return globals.has(n) ? globals.get(n) : 0n;
  }
  function ev(e, fr) {
    if (typeof e === 'bigint') return e;
    if (typeof e === 'string') return get(e, fr);
    const [h, ...a] = e;
    if (h === 'fn') return fnv(a[0], a[1], false);
    if (h === 'if') return ev(a[0], fr) !== 0n ? ev(a[1], fr) : ev(a[2], fr);
    const v = a.map((x) => ev(x, fr));
    if (h === 'call') {
      const [f, ...args] = v;
      if (typeof f !== 'object') return 0n;
      const nf = { loc: new Map(f.ps.map((p, i) => [p, args[i]])), gl: new Set() };
      if (!f.isProc) return ev(f.body, nf);
      try { run(f.body, nf); } catch (x) { if (x instanceof Ret) return x.v; throw x; }
      return 0n;
    }
    const fl = (x, y) => { let q = x / y; if ((x % y !== 0n) && ((x < 0n) !== (y < 0n))) q -= 1n; return q; };
    switch (h) {
      case '+': return v[0] + v[1]; case '-': return v[0] - v[1]; case '*': return v[0] * v[1];
      case '/': return v[1] === 0n ? 0n : fl(v[0], v[1]);
      case '%': return v[1] === 0n ? 0n : v[0] - fl(v[0], v[1]) * v[1];
      case '<': return v[0] < v[1] ? 1n : 0n; case '=': return v[0] === v[1] ? 1n : 0n;
    }
    throw new Error('op ' + h);
  }
  function assign(n, val, fr) { if (!fr || fr.gl.has(n)) globals.set(n, val); else fr.loc.set(n, val); }
  function run(body, fr) {
    for (const s of body) {
      switch (s.t) {
        case 'SET': assign(s.n, ev(s.e, fr), fr); break;
        case 'PRINT': out.push(ev(s.e, fr)); break;
        case 'RET': throw new Ret(ev(s.e, fr));
        case 'GLOBAL': fr.gl.add(s.n); break;
        case 'IF': run(ev(s.e, fr) !== 0n ? s.a : s.b, fr); break;
        case 'REPEAT': { const n = ev(s.e, fr); for (let i = 0n; i < n; i++) run(s.a, fr); break; }
        case 'DEF': assign(s.n, fnv(s.ps, s.a, true), fr); break;
      }
    }
  }
  run(prog, null);
  return out.map(String);
}
if (process.argv[2]) console.log(runArrow(readFileSync(process.argv[2], 'utf8')).join(' '));

File: tasks/bench-arrow-fn-values/fixtures/reference/arrow.py
import sys, re
def tok(s): return s.replace('(', ' ( ').replace(')', ' ) ').split()
def parse(t, i):
    if t[i] == '(':
        i += 1
        h = t[i]
        if h == 'fn':
            i += 1  # '('
            assert t[i] == '('
            i += 1; ps = []
            while t[i] != ')': ps.append(t[i]); i += 1
            i += 1
            body, i = parse(t, i)
            assert t[i] == ')'
            return ('fn', ps, body), i + 1
        i += 1; args = []
        while t[i] != ')':
            a, i = parse(t, i); args.append(a)
        return (h, args), i + 1
    if re.fullmatch(r'-?\d+', t[i]): return ('int', int(t[i])), i + 1
    return ('name', t[i]), i + 1
class Fn:
    def __init__(s, ps, body, stmts=False): s.ps, s.body, s.stmts = ps, body, stmts
class Ret(Exception):
    def __init__(s, v): s.v = v
def run(src):
    out = []; glob = {}
    lines = [l.strip() for l in src.split('\n')]
    lines = [l for l in lines if l and not l.startswith('#')]
    def block(i, stops):
        body = []
        while i < len(lines) and lines[i].split()[0] not in stops:
            st, i = stmt(i); body.append(st)
        return body, i
    def stmt(i):
        t = lines[i].split(None, 1); h = t[0]
        if h == 'SET':
            name, rest = t[1].split(None, 1); e, _ = parse(tok(rest), 0); return ('SET', name, e), i + 1
        if h == 'PRINT': e, _ = parse(tok(t[1]), 0); return ('PRINT', e), i + 1
        if h == 'RET': e, _ = parse(tok(t[1]), 0); return ('RET', e), i + 1
        if h == 'GLOBAL': return ('GLOBAL', t[1]), i + 1
        if h == 'IF':
            e, _ = parse(tok(t[1]), 0); b1, j = block(i + 1, ('ELSE', 'END')); b2 = []
            if lines[j].split()[0] == 'ELSE': b2, j = block(j + 1, ('END',))
            return ('IF', e, b1, b2), j + 1
        if h == 'REPEAT':
            e, _ = parse(tok(t[1]), 0); b, j = block(i + 1, ('END',)); return ('REPEAT', e, b), j + 1
        if h == 'DEF':
            parts = t[1].split(); b, j = block(i + 1, ('END',)); return ('DEF', parts[0], parts[1:], b), j + 1
        raise Exception('bad ' + lines[i])
    prog, _ = block(0, ())
    def read(n, fr):
        if fr is not None:
            if n in fr['g']: return glob.get(n, 0)
            if n in fr['l']: return fr['l'][n]
        return glob.get(n, 0)
    def ev(e, fr):
        k = e[0]
        if k == 'int': return e[1]
        if k == 'name': return read(e[1], fr)
        if k == 'fn': return Fn(e[1], e[2])
        if k == 'if':
            c = ev(e[1][0], fr)
            return ev(e[1][1], fr) if c != 0 else ev(e[1][2], fr)
        if k == 'call':
            f = ev(e[1][0], fr); vals = [ev(a, fr) for a in e[1][1:]]
            return callf(f, vals)
        v = [ev(a, fr) for a in e[1]]
        if k == '+': return v[0] + v[1]
        if k == '-': return v[0] - v[1]
        if k == '*': return v[0] * v[1]
        if k == '/': return 0 if v[1] == 0 else v[0] // v[1]
        if k == '%': return 0 if v[1] == 0 else v[0] % v[1]
        if k == '<': return int(v[0] < v[1])
        if k == '=': return int(v[0] == v[1])
        raise Exception('op ' + k)
    def callf(f, vals):
        if not isinstance(f, Fn): return 0
        fr = {'l': dict(zip(f.ps, vals)), 'g': set()}
        if not f.stmts: return ev(f.body, fr)
        try: exs(f.body, fr)
        except Ret as r: return r.v
        return 0
    def exs(body, fr):
        for s in body:
            k = s[0]
            if k == 'SET':
                v = ev(s[2], fr)
                if fr is None or s[1] in fr['g']: glob[s[1]] = v
                else: fr['l'][s[1]] = v
            elif k == 'PRINT': out.append(ev(s[1], fr))
            elif k == 'RET': raise Ret(ev(s[1], fr))
            elif k == 'GLOBAL': fr['g'].add(s[1])
            elif k == 'IF': exs(s[2] if ev(s[1], fr) != 0 else s[3], fr)
            elif k == 'REPEAT':
                for _ in range(ev(s[1], fr)): exs(s[2], fr)
            elif k == 'DEF':
                f = Fn(s[2], s[3], True)
                if fr is None or s[1] in fr['g']: glob[s[1]] = f
                else: fr['l'][s[1]] = f
    exs(prog, None)
    return out
if __name__ == '__main__':
    print(' '.join(str(x) for x in run(open(sys.argv[1]).read())))
